notify/firebase.go

```diff
@@ -13,42 +14,82 @@ import (
 
 var logger = logrus.New().WithField("module", "notify").WithField("service", "firebase")
 
-func SendPushBatch(messages []*messaging.Message) (*messaging.BatchResponse, error) {
+func isRelevantError(response *messaging.SendResponse) bool {
+	if !response.Success && response.Error != nil {
+		// Ignore https://stackoverflow.com/questions/58308835/using-firebase-for-notifications-getting-app-instance-has-been-unregistered
+		// Errors since they indicate that the user token is expired
+		if !strings.Contains(response.Error.Error(), "registration-token-not-registered") {
+			return true
+		}
+	}
+	return false
+}
+
+func SendPushBatch(messages []*messaging.Message) error {
 	credentialsPath := utils.Config.Notifications.FirebaseCredentialsPath
 	if credentialsPath == "" {
 		logger.Errorf("firebase credentials path not provided, disabling push notifications")
-		return nil, nil
+		return nil
 	}
 
 	ctx := context.Background()
 	opt := option.WithCredentialsFile(credentialsPath)
 	app, err := firebase.NewApp(context.Background(), nil, opt)
 	if err != nil {
 		logger.Errorf("error initializing app:  %v", err)
-		return nil, err
+		return err
 	}
 
 	client, err := app.Messaging(ctx)
 	if err != nil {
 		logger.Errorf("error initializing messaging: %v", err)
-		return nil, err
+		return err
 	}
 
-	result, err := client.SendAll(context.Background(), messages)
-	if err != nil {
-		logger.Errorf("error sending push notifications: %v", err)
-		return nil, err
+	var waitBeforeTryInSeconds = []time.Duration{0, 2, 4, 8, 16}
+	var resultSuccessCount, resultFailureCount int = 0, 0
+	var result *messaging.BatchResponse
+
+	currentMessages := messages
+	tries := 0
+	for _, s := range waitBeforeTryInSeconds {
+		time.Sleep(s * time.Second)
+		tries++
+
+		result, err = client.SendAll(context.Background(), currentMessages)
+		if err != nil {
+			logger.Errorf("error sending push notifications: %v", err)
+			return err
+		}
+
+		resultSuccessCount += result.SuccessCount
+		resultFailureCount += result.FailureCount
+
+		newMessages := make([]*messaging.Message, 0, resultFailureCount)
+		if resultFailureCount > 0 {
```

Comment:
Super unimportant but using `result.FailureCount` instead of `resultFailureCount` would make the code a bit easier to read since you don't have to look ahead for the decrement.